### DISCUSSION TITLE
Supporting Open Lineage in Unity Catalog

### DISCUSSION THREAD

**Discussion by mandy-chessell** [unitycatalog/unitycatalog#244] [Ideas]
**Is your feature request related to a problem? Please describe.**

Open Lineage (see https://openlineage.io/) defines a standard format for lineage events.  There are three types of events:

* Run Events - for the running of jobs
* Data Set Events - for publishing information about data sets
* Job Events - for publishing information about different types of jobs.

It is a standard that is gaining support in the industry.  Apache Spark and Apache Airflow can be set up to produce open lineage event.  Egeria both consumes and produces open lineage events.

**Describe the solution you would like**

This is a proposal to add support for open lineage events to Unity Catalog.  These events can be used for:

* Auditing activity around the use of Unity Catalog
* Connecting actions made by Unity catalog into lineage graphs that show the end-to-end flow of data and the operations acting upon it from the original operational sources to the lakehouse, through AI models and to its final consumers.

If we map the concepts from Open Lineage to unity catalog, this would involve:

* Sending a Data Set event each time a Table or Volume is created, updated, deleted
* Sending a Job Event each time a function is created, updated, deleted
* Sending a Run Event(START)  when a function is called
* Sending a Run Event (COMPLETE/ABORT/FAIL) when a function ends.

**Additional context**

Open Lineage events are issued to a HTTP endpoint.  The Open Lineage project provides a proxy back end that receives these events and forwards them through Kafka to one or more consumers.  In addition the Marquez project (https://marquezproject.ai/) supports the HTTP endpoint and ia able to store them and visualize lineage graphs.  Finally Egeria is able to receive open lineage events, either directly via HTTP, or via the Proxy back end and Kafka, and can:

* catalog jobs and data sets it detects along with their design lineage.
* save or forward the open lineage events, either as received or augmented with additional information.

Here is information about Egeria's support for open lineage.
*  https://egeria-project.org/features/lineage-management/overview/#the-open-lineage-standard

**Comment by MrPowers** [comment]
Thanks for raising this.  Let's move this over to a discussion so we can brainstorm the next steps as a community!

**Comment by mobuchowski** [comment]
From OpenLineage community perspective, it would be great to see better Unity Catalog integration.

**Comment by mandy-chessell** [comment]
I have been thinking more about what would be needed to change in UC in order to accommodate OpenLineage besides sending the events in the right format at the right times.

* OpenLineage needs a receiving HTTP endpoint running (eg Egeria, Marquez, the proxy back end, ...).  So that UC continues to have a seemless up and running experience, it should probably have OpenLineage turned off by default and have a big switch (eg environment variable) to supply it with the URL of the default receiving endpoint in order to activate it.

* Lineage events in general (whether the OpenLineage format is used or not) need to be selectively enabled and disabled because:
         * Lineage takes processing effort and UC is on the critical path for data retrieval.  The events are only generated when the metadata changes or a function is run, but performance still has to be a consideration.
         * Data in some schemas/tables etc is not interesting, so it should be possible to turn off lineage for these resources to avoid swamping the receiving endpoint with useless data.  Similarly, functions may not all be interesting for lineage.
         * Data in some schemas/tables may be extremely sensitive, but also needs lineage because it is critical to regulatory reporting, for example.   Even metadata about this data, and the functions being run against it, can be sensitive and so the lineage events may need to sent to a different endpoint.
         
All of these suggest that metadata element in UC would need to be extended to include an on/off switch and may a specific destination for the events for any lineage support, irrespective of wherher the OpenLineage format is used.

**Comment by rwforest** [comment]
@mandy-chessell  @mobuchowski FYI, for OSS Spark, Open Lineage already works quite well https://github.com/OpenLineage/OpenLineage. OL is also part of LF Data & AI. However, it's a bit challenging to get Databricks working with OL. Either Databricks opens source the lineage piece or integrate with OL and commit code there, is what I recommend. There isn't any secret on lineage; it is a very much-needed feature for governance.

**Comment by tvanderv92** [comment]
Hey everyone,

Just checking in to see if there are any updates on this. Open Lineage support in Unity Catalog would be really useful for lineage tracking, and I was wondering if this is on the roadmap or actively being considered.

Would love to hear if there’s been any progress or if there's anything the community can do to help push this forward.

Thanks!

**Reply by mandy-chessell** [reply]
There are two parts of consider:

* Unity Catalog producing open lineage events when it "does something". For example, when it is asked to run a function.  For an outside team to enable this,  we at least need Unity Catalog to provide an event, or plug point for a listener to trigger the open lineage event production.  We can then build a plugin to produce open lineage events as appropriate.  Databricks can have a proprietary plugin for their own lineage format.

* Unity Catalog receiving open lineage events and combining it with their own lineage to show what is happening either upstream or downstream of Unity Catalog.

The first is the most important to enable Unity Catalog to integrate into a broader enterprise governance space.  The second needs more of a discussion in the UC community to decide if the broader lineage mission is something that they want to take on.  I would suggest that with Egeria and Marquez covering much of this already, though their existing integration with Open Lineage, It may be a lot of work for little overall gain.
